### ISSUE TITLE
Mullvad blocks internet to macOS Virtualization.Framework/QEMU VMs

### ISSUE THREAD

**Issue by AwlsomeAlex** [mullvad/mullvadvpn-app#5213]
## Issue report

Operating system: macOS Sonoma (14.0.0 - 23A344); Apple Silicon

App version: 2023.5-beta2

### Issue description

Tried to connect to internet on Virtualization.Framework Linux VM. Connection wouldn't establish until Mullvad was disconnected.
Works fine in Parallels. Tried with two different Virtualization.Framework clients (UTM and VirtualBuddy) with same result.
Same result with UTM in QEMU mode as well.

**Comment by 666Danger-sudo** [comment]
Same on Debian with virtualbox.

**Comment by faern** [comment]
Thanks for reporting. We have created an internal ticket for this. For the Linux VM under macOS Virtualization.Framework specifically. I don't think it's the same issue stopping this from working on Debian + virtualbox. Those are almost definitely unrelated bugs.

**Comment by AwlsomeAlex** [comment]
@faern I will note, enabling "Local Network Sharing" fixes the issue for Parrallels, but not for UTM. I believe they use different IP ranges. My "Shared Network" guest looks to be using 192.168.64.x in UTM if that helps.

**Comment by faern** [comment]
I realize the issue does not state if Mullvad is running on the macOS host, or the Linux guest VM. That would be crucial to know. I have assumed it runs on the macOS host.

Can the host reach the VM and vice versa? Is the VM completely blocked out from the virtual network it's connected to, or is it just **forwarded** internet traffic that is blocked? Can you try pinging both ways?

**Comment by 666Danger-sudo** [comment]
On my side, it’s when mullvad is running on the host (debian) and there is a vm on virtualbox (no matter the os).Le 30 nov. 2023 à 12:12, Linus Färnstrand ***@***.***> a écrit :﻿
I realize the issue does not state if Mullvad is running on the macOS host, or the Linux guest VM. That would be crucial to know. I have assumed it runs on the macOS host.
Can the host reach the VM and vice versa? Is the VM completely blocked out from the virtual network it's connected to, or is it just forwarded internet traffic that is blocked? Can you try pinging both ways?

**Comment by conana** [comment]
Also having this issue.

Host OS: Sonoma 14.2.1 on M1 Pro 
Mullvad on host: 2023.6

VM: UTM 4.4.5
GuestOS: Fresh Win11 arm64 install

Guest OS can resolve DNS but not make any connections.

Mullvad enabled:
```
Pinging google.com [142.251.35.174] with 32 bytes of data:
Request timed out.
Request timed out.
Request timed out.
Request timed out.

Ping statistics for 142.251.35.174:
    Packets: Sent = 4, Received = 0, Lost = 4 (100% loss),
```

Mullvad disabled:
```
Pinging google.com [142.251.35.174] with 32 bytes of data:
Reply from 142.251.35.174: bytes=32 time=25ms TTL=113
Reply from 142.251.35.174: bytes=32 time=26ms TTL=113
Reply from 142.251.35.174: bytes=32 time=26ms TTL=113
Reply from 142.251.35.174: bytes=32 time=40ms TTL=113

Ping statistics for 142.251.35.174:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 25ms, Maximum = 40ms, Average = 29ms
```

Does not matter whether `Local Network Sharing` is enabled or not.
On the other hand, enabling a VPN with the wireguard.com App version 1.0.16 works just fine.

**Comment by conana** [comment]
Found a sufficient workaround (kudos to Mullvad support). Changing QEMU network mode for the VM from the default `Shared Network` to `Emulated VLAN` (did not change any other setting) was enough to allow the guest to access Internet hosts. Traffic is routing correctly via the VPN.

**Comment by faern** [comment]
Yeah, VMs is hard! Because it depends on how the hypervisor implements networking. Mullvad only looks at and cares abount Layer 3 traffic (IP). If the VM networking is done on Layer 2 (Ethernet) it becomes transparent to our app's firewall rules and can cause various problems.

**Comment by Dmusulas** [comment]
Are there any updates on this issue? I have stumbled upon this problem and no matter which network mode I select (apart from Bridged) MullvadVPN blocks the connection. Only way I can use VMs with MullvadVPN installed is by disabling it which is not ideal...

**Comment by Bendodroid** [comment]
I have OP's problem with App version 2025.5, does not seem to be fixed.

**Comment by yveszoundi** [comment]
**tl;dr** Maybe add `nameserver 10.64.0.1` to your DHCP configuration (if applicable) and remove any other DNS entries.

I had similar problems recently with a [Guix system](https://guix.gnu.org/en/download/) virtual machine running with `QEMU` on a macOS host (Mullvad VPN running on host, latest macOS on x86_64). 

- Enabling "Local Network Sharing" in the Mullvad VPN app had zero effect
- Fiddling with QEMU network parameters didn't change anything
- Strangely, all my other virtual machines (but `Guix`) had no issues with internet connectivity (Fedora, Alpine, etc.)

#### Potential solution
What worked for me was the following

- Ensure that in my DHCP configuration I didn't have any custom `nameserver` entries such as [OpenDNS](https://www.opendns.com/setupguide/) or [Google DNS servers](https://developers.google.com/speed/public-dns/docs/using).
- Add a new `nameserver` entry to `/etc/resolv.conf.head` with `nameserver 10.64.0.1`
  - See also https://discourse.nixos.org/t/connected-to-mullvadvpn-but-no-internet-connection/35803/11
  - See also https://mullvad.net/en/help/running-wireguard-router (additional reference to `10.64.0.1`)
- The only other `nameserver` entry that I have at the bottom of `/etc/resolv.conf` is auto-generated by `dhcpcd`

**Comment by simpleapps-public** [comment]
I'm also running into this issue trying to use the combination of the Mullvad app and the macOS Virtualization.Framework (with a wrapper like UTM or VirtualBuddy and a "shared" network device). I've found a solution using `pf` to route the packets.

In the macOS host, edit `/etc/pf.conf` and add this before any existing "anchor" entries:

```
nat on utun0 from 192.168.64.0/24 to any -> (utun0)
pass in on bridge100 inet from 192.168.64.0/24 to any keep state
pass out on utun0 inet from 192.168.64.0/24 to any keep state
```

Your exact values may be different. utun0 is Mullvad's  interface. bridge100 is the interface created by UTM. 192.168.64.0/24 is the subnet range UTM will assign addresses on.

Then run `sudo pfctl -F all -f /etc/pf.conf` Everything should now work. However this solution is quite fiddly and pfctl needs to be rerun anytime the Mullvad app reconnects. 

I found this solution here and the author has more insights and suggestions (such as lowering the VM guest MTU):  https://github.com/utmapp/UTM/issues/3238#issuecomment-3314689110

Perhaps these insights can aid the Mullvad app developers.
